fenix/app/src/main/java/org/mozilla/fenix/library/history/state/HistoryNavigationMiddleware.kt

```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.library.history.state
+
+import androidx.navigation.NavController
+import androidx.navigation.NavOptions
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.R
+import org.mozilla.fenix.library.history.History
+import org.mozilla.fenix.library.history.HistoryFragmentAction
+import org.mozilla.fenix.library.history.HistoryFragmentDirections
+import org.mozilla.fenix.library.history.HistoryFragmentState
+
+/**
+ * A [Middleware] for initiating navigation events based on [HistoryFragmentAction]s that are
+ * dispatched to the [HistoryFragmentStore].
+ *
+ * @property navController [NavController] for handling navigation events
+ * @property openToBrowser Callback to open history items in a browser window.
+ */
+class HistoryNavigationMiddleware(
+    private val navController: NavController,
+    private val openToBrowser: (item: History.Regular) -> Unit,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main),
+) : Middleware<HistoryFragmentState, HistoryFragmentAction> {
+    override fun invoke(
+        context: MiddlewareContext<HistoryFragmentState, HistoryFragmentAction>,
+        next: (HistoryFragmentAction) -> Unit,
+        action: HistoryFragmentAction,
+    ) {
+        next(action)
+        scope.launch {
+            when (action) {
+                is HistoryFragmentAction.HistoryItemClicked -> {
+                    if (context.state.mode.selectedItems.isEmpty()) {
+                        when (val item = action.item) {
+                            is History.Regular -> openToBrowser(item)
```

Comment:
rather than a lambda, should this be firing off a new Action instead for the telemetry and `openToBrowserAndLoad` logic?

Reply:
I do like the idea of firing an Action to the Browser Store to handle the page opening logic. I would prefer to do this in a follow-up though and continue to re-use the logic as it is still shared with the controller until the flag is removed. How does that sound?

The telemetry is already captured by the telemetry middleware, though I'm now realizing that means it will be recorded twice and so the flag should probably be used to prevent that.

Reply:
https://bugzilla.mozilla.org/show_bug.cgi?id=1845409